# Practice Session 04: Basket analysis

Author: <font color="blue">Leticia Martin Cabrera</font>

E-mail: <font color="blue">leticia.martin03@estudiant.upf.edu</font>

Date: <font color="blue">03/11/2021</font>

In [203]:
pip install apyori

Note: you may need to restart the kernel to use updated packages.


In [204]:
import numpy as np  
import matplotlib.pyplot as plt  
import pandas as pd  
from apyori import apriori

# 1. Playing with apyori

In [205]:
#This example contains 20 samples of the social media of the people use.
social_media = [
    ['Facebook', 'Instagram', 'Twitter', 'Linkedin', 'WhatsApp'],
    ['Facebook', 'Instagram', 'WhatsApp'],
    ['Instagram', 'Twitter', 'WhatsApp'],
    ['Linkedin','WhatsApp'],
    ['Facebook','Twitter','WhatsApp'],
    ['Facebook', 'Linkedin'],
    ['Facebook', 'Twitter', 'Linkedin'],
    ['Facebook','Linkedin','WhatsApp'],
    ['Instagram','Twitter', 'Linkedin'],
    ['WhatsApp'],
    ['Instagram','WhatsApp'],
    ['Facebook', 'Twitter'],
    ['Facebook','Instagram', 'Twitter','WhatsApp'],
    ['Facebook','WhatsApp'],
    ['Twitter','WhatsApp'],
    ['Instagram','Twitter' ],
    ['Instagram','Twitter','Linkedin','WhatsApp'],
    ['Facebook'],
    ['Facebook','Twitter','Linkedin','WhatsApp'],
    ['Facebook','Instagram','Linkedin','WhatsApp']
]

results = list(apriori(social_media, min_support=0.2, min_confidence=0.75, min_lift=1.0))
results

[RelationRecord(items=frozenset({'WhatsApp', 'Instagram'}), support=0.35, ordered_statistics=[OrderedStatistic(items_base=frozenset({'Instagram'}), items_add=frozenset({'WhatsApp'}), confidence=0.7777777777777777, lift=1.111111111111111)]),
 RelationRecord(items=frozenset({'Facebook', 'WhatsApp', 'Instagram'}), support=0.2, ordered_statistics=[OrderedStatistic(items_base=frozenset({'Facebook', 'Instagram'}), items_add=frozenset({'WhatsApp'}), confidence=1.0, lift=1.4285714285714286)])]

In [206]:
# Leave this code as-is

def print_apyori_output (association_results):
    for relation_record in association_results:
        itemset = list(relation_record.items)
        
        # Consider only itemsets of two elements
        if len(itemset) > 1: 
        
            print("Rules involving itemset %s" % itemset)
            support = relation_record.support

            for rules in relation_record.ordered_statistics:
                antecedent = list(rules.items_base)
                consequent = list(rules.items_add)
                confidence = rules.confidence
                lift = rules.lift

                print("%s => %s (support=%.2f, confidence=%.2f, lift=%.2f)" %
                      (antecedent, consequent, support, confidence, lift))
            print()

In [207]:
print_apyori_output(results)

Rules involving itemset ['WhatsApp', 'Instagram']
['Instagram'] => ['WhatsApp'] (support=0.35, confidence=0.78, lift=1.11)

Rules involving itemset ['Facebook', 'WhatsApp', 'Instagram']
['Facebook', 'Instagram'] => ['WhatsApp'] (support=0.20, confidence=1.00, lift=1.43)



I have obtained 2 different rules involving the itemset **['WhatsApp', 'Instagram']** and the itemset **['Facebook', 'WhatsApp', 'Instagram']**.

In the first rule we have obtained a support of 0.35, a confidence of 0.78 and a lift of 1.11. Now, we explain how calculate it.

In [208]:
A = 'Instagram' #item A
B = 'WhatsApp'  #item B

#SUPPORT
"""To calculate the support, we divide the number of transactions containing Instagram an WhatsApp 
by the total number of transactions"""

transactions_A_B = 0 #total number of transactions containing Instagram and WhatsApp
total_transactions = len(social_media)
for i in social_media:
    if (A in i)  and (B in i): #transaction i contains Instagram and WhatsApp
        transactions_A_B += 1 

support = transactions_A_B/total_transactions
print("The support of Instagram and WhatsApp is", support)


#CONFIDENCE
"""In order to calculate the level of confidence, find the number of transactions in which Instagram and WhatsApp 
are bought together, divided by the number of transactions in which Instagram is bought."""
transactions_A = 0 #total number of transactions containing Instagram
for i in social_media:
    if A in i: #transaction i contains Instagram 
        transactions_A += 1 
confidence = transactions_A_B/transactions_A
print("The confidence of Instagram and WhatsApp is", "{:.2f}".format(confidence))


#LIFT
"""Lift(Instagram –> WhatsApp) can be calculated by dividing Confidence(Instagram -> WhatsApp) by Support(WhatsApp)"""

#calculate the support of WhatsApp
transactions_B = 0 #total number of transactions containing WhatsApp
for i in social_media:
    if B in i: #transaction i contains WhatsApp
        transactions_B += 1
support_B = transactions_B/total_transactions #support of WhatsApp

lift = confidence/support_B #lift (Instagram –> WhatsApp)
print("The lift of Instagram and WhatsApp is", "{:.2f}".format(lift))

The support of Instagram and WhatsApp is 0.35
The confidence of Instagram and WhatsApp is 0.78
The lift of Instagram and WhatsApp is 1.11


In the second rule we have obtained a support of 0.2, a confidence of 1.00 and a lift of 1.43. We calculate it as the same way of above.

In [209]:
A = 'Facebook'  #item A
B = 'Instagram' #item B
C = 'WhatsApp'  #item c
 
#SUPPORT
"""To calculate the support, we divide the number of transactions containing Facebook,Instagram an WhatsApp 
by the total number of transactions"""

transactions_A_B_C = 0 #total number of transactions containing Instagram and WhatsApp
total_transactions = len(social_media)
for i in social_media:
    if (A in i)  and (B in i) and (C in i): #transaction i contains Facebook, Instagram and WhatsApp
        transactions_A_B_C += 1 

support = transactions_A_B_C/total_transactions
print("The support of Facebook,Instagram and WhatsApp is", support)


#CONFIDENCE
"""In order to calculate the level of confidence, find the number of transactions in which Facebook,Instagram and WhatsApp 
are bought together, divided by the number of transactions in which Facebook and Instagram  is bought."""
transactions_A_B = 0 #total number of transactions containing Facebook and Instagram 
for i in social_media:
    if (A in i)  and (B in i): #transaction i contains Instagram 
        transactions_A_B+= 1 
confidence = transactions_A_B_C/transactions_A_B
print("The confidence of Facebook,Instagram and WhatsApp is", "{:.2f}".format(confidence))


#LIFT
"""Lift(Facebook and Instagram –> WhatsApp) can be calculated by dividing 
Confidence(Facebook and Instagram -> WhatsApp) by Support(WhatsApp)"""

#calculate the support of WhatsApp
transactions_C = 0 #total number of transactions containing WhatsApp
for i in social_media:
    if C in i: #transaction i contains WhatsApp 
        transactions_C += 1
support_C = transactions_C/total_transactions #support of WhatsApp

lift = confidence/support_C #lift (Instagram –> WhatsApp)
print("The lift of Instagram and WhatsApp is", "{:.2f}".format(lift))

The support of Facebook,Instagram and WhatsApp is 0.2
The confidence of Facebook,Instagram and WhatsApp is 1.00
The lift of Instagram and WhatsApp is 1.43


# 2. Load and prepare the services purchased dataset

In [210]:
INPUT_FILENAME = "data/services_purchased.csv"

In [211]:
dataset = pd.read_csv(INPUT_FILENAME, sep=",")
dataset.head()

ID_customer  WEBHOSTING  OFFICESUITE  SECURITY  CLOUD_IAAS  CLOUD_PAAS  \
0            0           0            0         1           0           0   
1            1           0            1         1           0           0   
2            2           1            0         1           0           0   
3            3           0            0         1           0           0   
4            4           1            1         1           0           0   

   CONTENTMGM  CHATBOT  ADVERTISING  
0           0        0            0  
1           0        0            0  
2           1        0            0  
3           0        0            0  
4           1        0            0

In [212]:
for service in dataset.iloc[:,1:,]:
    print(service,":", np.sum(dataset[service]))
    

WEBHOSTING : 274
OFFICESUITE : 176
SECURITY : 608
CLOUD_IAAS : 67
CLOUD_PAAS : 6
CONTENTMGM : 152
CHATBOT : 0
ADVERTISING : 9


In [213]:
dataset= dataset.drop(labels="ID_customer", axis=1)
shape= dataset.shape
print("The number of columns after removing the ID_customer column is:", shape[1])

The number of columns after removing the ID_customer column is: 8


In [214]:
dataset =dataset.loc[(dataset!=0).any(axis=1)]
shape= dataset.shape
print("The number of rows after removing all customers that have not purchased any service is:", shape[0])

The number of rows after removing all customers that have not purchased any service is: 753


WEBHOSTING  OFFICESUITE  SECURITY  CLOUD_IAAS  CLOUD_PAAS  CONTENTMGM  \
0             0            0         1           0           0           0   
1             0            1         1           0           0           0   
2             1            0         1           0           0           1   
3             0            0         1           0           0           0   
4             1            1         1           0           0           1   
..          ...          ...       ...         ...         ...         ...   
993           0            0         1           0           0           0   
994           0            0         1           0           0           0   
995           0            0         1           0           0           0   
998           1            0         1           0           0           0   
999           1            0         0           0           0           0   

     CHATBOT  ADVERTISING  
0          0            0  
1          0            0  
2          0            0  
3          0            0  
4          0            0  
..       ...          ...  
993        0            0  
994        0            0  
995        0            0  
998        0            0  
999        0            0  

[753 rows x 8 columns]

In [215]:
transactions = []
for recordnum, record in dataset.iterrows():
    transaction = [] #elements of the transactions varieble 
    for column in dataset.columns:
        if dataset[column][recordnum]:
            transaction.append(column) #adding the service to a new element in the transactions list 
    transactions.append(transaction) #adding new elements to the transactions variable

transactions[:5] #checking the first five elements 


[['SECURITY'],
 ['OFFICESUITE', 'SECURITY'],
 ['WEBHOSTING', 'SECURITY', 'CONTENTMGM'],
 ['SECURITY'],
 ['WEBHOSTING', 'OFFICESUITE', 'SECURITY', 'CONTENTMGM']]

# 3. Run the Apriori algorithm

In [216]:
#High-Support rules
results1 = apriori(transactions, min_support= 0.25, min_confidence = 0.25, min_lift = 0.85)
print_apyori_output(results1)


Rules involving itemset ['SECURITY', 'WEBHOSTING']
[] => ['SECURITY', 'WEBHOSTING'] (support=0.25, confidence=0.25, lift=1.00)
['SECURITY'] => ['WEBHOSTING'] (support=0.25, confidence=0.31, lift=0.86)
['WEBHOSTING'] => ['SECURITY'] (support=0.25, confidence=0.70, lift=0.86)



The highest support value to get more than 1 rule is 0.25. I have obtained 3 rules involving the itemset ['WEBHOSTING', 'SECURITY'].  

The 1st rule [] => ['WEBHOSTING', 'SECURITY'] has a lift of 1.00 so it indicates that there is no association between [] and ['WEBHOSTING', 'SECURITY']. 

The 2nd rule ['SECURITY'] => ['WEBHOSTING']:
    - It has  a lift of 0.86 which indicates that these two products are unlikely to be bought together. 
    - It has confidence of 0.35 which means that it is unlikely that a Web Hosting product will be also purchased if Security product is purchased.

The 3rd rule ['WEBHOSTING'] => ['SECURITY']:
    - It has the same lif as the 2nd rule so these two products are unlikely to be bought together
    - It has a higher confidence, which means that if a Web hosting product is purchased, it is more likely that a Security product will also be purchased.

In [217]:
#High-lift rules
results2 = apriori(transactions, min_support= 0.15, min_confidence= 0.2, min_lift = 1.12)
print_apyori_output(results2)


Rules involving itemset ['SECURITY', 'CONTENTMGM']
['CONTENTMGM'] => ['SECURITY'] (support=0.18, confidence=0.91, lift=1.12)
['SECURITY'] => ['CONTENTMGM'] (support=0.18, confidence=0.23, lift=1.12)



With a vule of lift of 1.12, I could get more than 1 rule for the itemsets ['SECURITY', 'CONTENTMGM']:

The 1st rule ['CONTENTMGM'] => ['SECURITY']:
    - It has a confidence of 0.91 means that if the Content Management product is purchased, there is a 91% probability that Security will also be purchased.
    - It has a lift of 1.12 indicates that Security and Content management products have a higher probability to be purchased together.

The 2nd rule ['SECURITY'] => ['CONTENTMGM']:
    - It has the same lift as the 1st rule so Security and Content management products have a higher probability to be purchased together.
    - It has a lower confidence, showing to us that if the securty product is purchased, there is a few probabilities, 23%,  that Content Management product will also be purchased



In [223]:
dataset1 = dataset.loc[dataset.CONTENTMGM != 0] #new dataset containing the customers that purchase CONTENTMGM
dataset1 = dataset1.drop("CONTENTMGM",axis=1)   #removing the CONTENTMGM column of our new dataset

#create the "transactions" list from the dataset1
transactions1 = []
for recordnum, record in dataset1.iterrows():
    transaction = [] #elements of the transactions varieble 
    for column in dataset1.columns:
        if dataset1[column][recordnum]:
            transaction.append(column) #adding the service to a new element in the transactions list 
    transactions1.append(transaction) #adding new elements to the transactions variable

results3 = apriori(transactions1)
print_apyori_output(results3)

Rules involving itemset ['SECURITY', 'OFFICESUITE']
[] => ['SECURITY', 'OFFICESUITE'] (support=0.24, confidence=0.24, lift=1.00)
['OFFICESUITE'] => ['SECURITY'] (support=0.24, confidence=0.82, lift=0.90)
['SECURITY'] => ['OFFICESUITE'] (support=0.24, confidence=0.26, lift=0.90)

Rules involving itemset ['SECURITY', 'WEBHOSTING']
[] => ['SECURITY', 'WEBHOSTING'] (support=0.34, confidence=0.34, lift=1.00)
['SECURITY'] => ['WEBHOSTING'] (support=0.34, confidence=0.37, lift=1.02)
['WEBHOSTING'] => ['SECURITY'] (support=0.34, confidence=0.93, lift=1.02)



We can reject the rules involving itemset ['SECURITY', 'OFFICESUITE'] since these have a low confidence value( 1st and 3rd rule) and also the second rule has a lif value less than 1 so it is unlikely that office suite and security products are purchased together. 

Observing the rule ['WEBHOSTING'] => ['SECURITY'] we can suppose that the customers of the product Content MGM will buy Security services if they buy Web Hosting services. Since the lift value is greater than 1 and the confidence value is nearly 1. There is a 93% probability that Security service will also be purchased .

In [225]:
dataset2 = dataset.loc[dataset.WEBHOSTING != 0] #new dataset containing the customers that purchase WEBHOSTING
dataset2 = dataset2.drop("WEBHOSTING",axis=1)   #removing the WEBHOSTING column of our new dataset

#create the "transactions" list from the dataset1
transactions2 = []
for recordnum, record in dataset2.iterrows():
    transaction = [] #elements of the transactions varieble 
    for column in dataset2.columns:
        if dataset2[column][recordnum]:
            transaction.append(column) #adding the service to a new element in the transactions list 
    transactions2.append(transaction) #adding new elements to the transactions variable

results4 = apriori(transactions2, min_lift = 1.02)
print_apyori_output(results4)

Rules involving itemset ['SECURITY', 'CONTENTMGM']
['CONTENTMGM'] => ['SECURITY'] (support=0.19, confidence=0.93, lift=1.33)
['SECURITY'] => ['CONTENTMGM'] (support=0.19, confidence=0.27, lift=1.33)

Rules involving itemset ['SECURITY', 'OFFICESUITE']
['OFFICESUITE'] => ['SECURITY'] (support=0.13, confidence=0.71, lift=1.02)
['SECURITY'] => ['OFFICESUITE'] (support=0.13, confidence=0.18, lift=1.02)



We can reject the rules with low confidence value: ['SECURITY'] => ['CONTENTMGM']  and ['SECURITY'] => ['OFFICESUITE'].

Observing the rule ['CONTENTMGM'] => ['SECURITY'] we can suppose that the customers of the product Web Hosting will buy Security services if they buy content management solution. Since the lift value is greater than 1 and the confidence value is nearly 1. There is a 93% probability that Security service will also be purchased.

Also, observing the rule ['OFFICESUITE'] => ['SECURITY'] we can suppose, with less certainty than with the previous rule, that the customers of the product Web Hosting will buy security services if they buy office suite services. Since the lift value is greater than 1 and the confidence value is high. There is a 71% probability that Security service will also be purchased.

<font size="+2" color="#003300">I hereby declare that, except for the code provided by the course instructors, all of my code, report, and figures were produced by myself.</font>